rpcs3/Emu/RSX/VK/VKProgramPipeline.cpp

```diff
@@ -69,165 +142,466 @@ namespace vk
 		{
-			for (const auto &uniform : uniforms[program_input_type::input_type_texture])
+			for (unsigned i = 0; i < ::size32(m_sets); ++i)
 			{
-				if (uniform.name == uniform_name)
+				const auto& type_arr = m_sets[i].m_inputs[type];
+				const auto result = std::find_if(type_arr.cbegin(), type_arr.cend(), [&](const auto& u)
 				{
-					set.push(image_descriptor, type, uniform.location);
-					attribute_location_mask |= (1ull << uniform.location);
-					return;
+					return u.domain == domain && u.name == uniform_name;
+				});
+
+				if (result != type_arr.end())
+				{
+					return { i, result->location };
 				}
 			}
 
-			rsx_log.notice("texture not found in program: %s", uniform_name.c_str());
+			return { umax, umax };
 		}
 
-		void program::bind_uniform(const VkDescriptorImageInfo & image_descriptor, int texture_unit, ::glsl::program_domain domain, vk::descriptor_set &set, bool is_stencil_mirror)
+		void program::bind_uniform(const VkDescriptorImageInfo& image_descriptor, u32 set_id, u32 binding_point)
 		{
-			ensure(domain != ::glsl::program_domain::glsl_compute_program);
+			if (m_sets[set_id].m_descriptor_slots[binding_point] == image_descriptor)
+			{
+				return;
+			}
+
+			m_sets[set_id].notify_descriptor_slot_updated(binding_point, image_descriptor);
+		}
 
-			u32 binding;
-			if (domain == ::glsl::program_domain::glsl_fragment_program)
+		void program::bind_uniform(const VkDescriptorBufferInfo &buffer_descriptor, u32 set_id, u32 binding_point)
+		{
+			if (m_sets[set_id].m_descriptor_slots[binding_point] == buffer_descriptor)
 			{
-				binding = (is_stencil_mirror) ? fs_texture_mirror_bindings[texture_unit] : fs_texture_bindings[texture_unit];
+				return;
 			}
-			else
+
+			m_sets[set_id].notify_descriptor_slot_updated(binding_point, buffer_descriptor);
+		}
+
+		void program::bind_uniform(const VkBufferView &buffer_view, u32 set_id, u32 binding_point)
+		{
+			if (m_sets[set_id].m_descriptor_slots[binding_point] == buffer_view)
 			{
-				binding = vs_texture_bindings[texture_unit];
+				return;
 			}
 
-			if (binding != ~0u)
+			m_sets[set_id].notify_descriptor_slot_updated(binding_point, buffer_view);
+		}
+
+		void program::bind_uniform_array(const VkDescriptorImageInfo* image_descriptors, VkDescriptorType type, int count, u32 set_id, u32 binding_point)
+		{
+			auto& set = m_sets[set_id];
+			for (int i = 0; i < count; ++i)
+			{
+				if (set.m_descriptor_slots[binding_point + i] != image_descriptors[i])
+				{
+					set.notify_descriptor_slot_updated(binding_point + i, image_descriptors[i]);
+				}
+			}
+		}
+
+		void program::create_pipeline_layout()
+		{
+			ensure(!m_linked);
+			ensure(m_pipeline_layout == VK_NULL_HANDLE);
+
+			rsx::simple_array<VkPushConstantRange> push_constants{};
+			rsx::simple_array<VkDescriptorSetLayout> set_layouts{};
+
+			for (auto& set : m_sets)
+			{
+				if (!set.m_device)
+				{
+					break;
```

Comment:
Why break?

Reply:
Because it usually just implies there's nothing else after this set. Though this isn't performance-critical code so maybe continue is more readable and also removes a foot-gun in the layouts change in future.